Missing values in r (#121)
arcus/education_modules: r_missing_values/r_missing_values.md

<!--

author:   Rose Hartman
email:    hartmanr1@chop.edu
version:  0.0.1
module_template_version: 2.0.0
language: en
narrator: UK English Female
title: Missing Values in R
comment:  A practical demonstration of how missing values show up in R and how to deal with them. Note that this module does **not** cover statistical approaches for handling missing data, but instead focuses on the code you need to find, work with, and assign missing values in R.

> long_description: This is a beginners guide to handling missing values in R. It covers what `NA` values are and how to check for them, how to mark values as missing, how to work around missing values in your analysis, and how to filter your data to remove missingness. This module is appropriate for learners who feel comfortable with R basics and are ready to get into the realities of data analysis, like dealing with missing values.

Reviewer: ```suggestion
long_description: This is a beginner's guide to handling missing values in R. It covers what `NA` values are and how to check for them, how to mark values as missing, how to work around missing values in your analysis, and how to filter your data to remove missingness. This module is appropriate for learners who feel comfortable with R basics and are ready to get into the realities of data analysis, like dealing with missing values.
```

@end

script:  https://code.jquery.com/jquery-3.6.0.slim.min.js

@gifPreload
<script>
(function($) {

  // Get the .gif images from the "data-alt".
	var getGif = function() {
		var gif = [];
		$('img').each(function() {
			var data = $(this).data('alt');
			gif.push(data);
		});
		return gif;
	}

	var gif = getGif();

	// Preload all the gif images.
	var image = [];

	$.each(gif, function(index) {
		image[index]     = new Image();
		image[index].src = gif[index];
	});

	// Change the image to .gif when clicked and vice versa.
	$('figure').on('click', function() {

		var $this   = $(this),
				$index  = $this.index(),

				$img    = $this.children('img'),
				$imgSrc = $img.attr('src'),
				$imgAlt = $img.attr('data-alt'),
				$imgExt = $imgAlt.split('.');

		if($imgExt[1] === 'gif') {
			$img.attr('src', $img.data('alt')).attr('data-alt', $imgSrc);
		} else {
			$img.attr('src', $imgAlt).attr('data-alt', $img.data('alt'));
		}

		// Add play class to help with the styling.
		$this.toggleClass('play');

	});

})(jQuery);
</script>
@end

link:  https://chop-dbhi-arcus-education-website-assets.s3.amazonaws.com/css/styles.css
script: https://kit.fontawesome.com/83b2343bd4.js

-->

> # @title

Reviewer: ```suggestion
Missing Values in R
```
Unfortunately, using `@title` causes the left TOC to break, IDK why...

Author: Ah, thank you! I keep forgetting that there's a good reason we don't use @title as a header...

Although techniques for analyzing and correcting for different missingness patterns is outside the scope of this module,
one of the first steps in any of those techniques is checking how much missing data you have and on which variables, which we will cover here.

<div class="learnmore">
For an excellent introduction to different types of missing data, read Rubin's classic paper [Inference and Missing Data](http://math.wsu.edu/faculty/xchen/stat115/lectureNotes3/Rubin%20Inference%20and%20Missing%20Data.pdf).
</div>

Even if you won't be doing any formal techniques to correct for missing data, it's still important to know what missing data you have because it will impact your effective sample size.
Moreover, checking your missingness is one of the key steps you can take to understand your data quality and any cleaning steps that may be necessary.

So how can you tell when you have missing data in R?

Different statistical software mark missing values in different ways.
In R, missing values are marked with `NA`, which is short for "Not Available".

<div class = "learnmore">

> Technically, there are actually four different kinds of `NA`, one for each of the different data types in R. To learn more, read about [missing values in the free online book R for Data Science](https://r4ds.had.co.nz/vectors.html?q=missing#missing-values-4)

Reviewer: ```suggestion
Technically, there are actually four different kinds of `NA`, one for each of the different data types in R. To learn more, read about [missing values in the free online book R for Data Science](https://r4ds.had.co.nz/vectors.html?q=missing#missing-values-4).
```

Such techniques are outside the scope of this module, but there are lots of excellent papers and tutorials available with instructions.  
</div>
****

## Assigning values to missing

As part of your data exploration, you may notice values for some variables that are impossible or extremely unlikely.

For example, imagine you have a dataset called `df` with a variable called `rating`.
You know it's on a scale from 1-5, so respondents weren't able to rate it as anything outside of those options.
And yet, when you read in the data you see there are a handful of -99 scores for this variable.
Depending on the data collection mechanism, these may be the result of typos during data entry or a systematic way of marking invalid or missing responses.
Either way, you don't want to treat those -99 scores as actual responses to the question (imagine how they would distort the estimate of the mean for that variable!); instead, you can mark them missing so they'll be excluded from analysis.

Here's some code to assign the -99 values in our pretend dataset to missing:

```r
mutate(df, rating = if_else(rating == -99, NA, rating))
```

Let's break that down.

> We're using a `mutate` command to change the variable, and we're setting it to be conditional by using `if_else`.

Reviewer: ```suggestion
We're using a `mutate` command to change ("mutate") a variable (a column of data), and we're setting what the new value of this column should be by using a special conditional function, the `if_else` function.
```

> The `if_else` command begins with a conditional test, in this case `if_else(rating == -99, NA, rating)`, and for each value in `rating`, it will run the test and return either `TRUE` or `FALSE`.

Reviewer: ```suggestion
The `if_else` command begins with a conditional test, in this case `rating == -99`.  Remember that the double equals sign is a comparison -- it's **asking** whether `rating` is equal to -99, while a single equals sign is a "setter", and it will try to **make** `rating` equal to -99.  For each value in `rating`, R will run the test `rating = -99` and return either `TRUE` or `FALSE`.
```

Author: Couldn't batch this with the other suggestions because it was two conflicting suggestions for line 291, but I did implement it in a subsequent commit. Much better wording than what I had --- thank you!

> The `if_else` command begins with a conditional test, in this case `if_else(rating == -99, NA, rating)`, and for each value in `rating`, it will run the test and return either `TRUE` or `FALSE`.

Reviewer: ```suggestion
The `if_else` command is what's called a ternary operator, and it has three parts:

 * a conditional test
 * a value to use if the test returns `TRUE`
 * a value to use if the test returns `FALSE`.
 
 Our `if_else` statement begins with a conditional test, in this case `if_else(rating == -99, NA, rating)`, and for each value in `rating`, it will run the test and return either `TRUE` or `FALSE`.
```
